### PR TITLE
testdrive: allow deprecating statements based on version

### PR DESCRIPTION
**Pull request by sploiselle**
Attempting to atomize #15813 into more independently mergeable bits; #15813 has the commits in situ and I am glad to explain each commit in more detail.

@benesch just pinging you for a backward compatibility sign off or if you have a more elegant suggestion for the API.

### Motivation

  * This PR adds a known-desirable feature, which will form the basis of queryable remap shards.

ec41bbb0e629efeb19c48e34a31e865b9ad5096b is necessary because we changing the output of the output of `SHOW CREATE...`; we can retain the tests' output from the "current" version during upgrade tests, but then test that the output changes at some specified version. With the current way the tests are structured, there is no means to permit the output to change version-by-version.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Comment by benesch** [comment]
> With the current way the tests are structured, there is no means to permit the output to change version-by-version.

There actually _is_ a way. I don't necessarily recommend it, but here it is:

  * https://github.com/MaterializeInc/materialize/blob/bdbdfdf148c35ee03b0015ae39f09507dda0dd88/test/upgrade/create-in-v0.27.0-kafka-source.td#L10-L13
  * https://github.com/MaterializeInc/materialize/blob/main/test/upgrade/create-in-v0.39.0-kafka-source.td#L10-L13
  * https://github.com/MaterializeInc/materialize/blob/main/test/upgrade/check-from-v0.27.0-kafka-source.td#L39-L40

tl;dr write to a SQL table with the desired output; read back the desired output in the check script.

**Comment by benesch** [comment]
Looking at #15813, I have a real recommendation! You might want to just conditionally skip the end of that file using `$ skip-if`? [Docs.](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/testdrive.md#conditionally-skipping-the-rest-of-a-td-file)

```
$ skip-if
SELECT mz_version_num() < 3900;
```

You'll have to figure out what the right `mz_version_num` value is. I forget how it works off the top of my head.

----

Nothing terribly wrong with what you've got here, but it's a funky API. Would feel better about it if `> ` commands could accept options directly but, sigh, they can't. (The legacy upgrade tests are going away at some point, so I'm rather loathe to add testdrive features that are just to facilitate these kinda wonky legacy upgrade tests.)

**Comment by sploiselle** [comment]
@benesch great--will proceed with `skip-if`. wasn't aware of `mz_version_num`
